aten/src/ATen/core/jit_type.h

```diff
@@ -97,6 +97,78 @@ struct SingleElementType : public Type {
   TypePtr elem;
 };
 
+struct UnionType;
+using UnionTypePtr = std::shared_ptr<UnionType>;
+struct TORCH_API UnionType : public Type {
+
+  friend struct Type;
+
+  static const TypeKind Kind = TypeKind::UnionType;
+
+  bool isSubtypeOfExt(const TypePtr& rhs_, std::ostream* why_not) const;
+
+  std::string str() const;
+
+  std::vector<TypePtr> present_types() const;
+
+  static UnionTypePtr create(
+      std::vector<TypePtr> types) {
+    return UnionTypePtr(new UnionType(
+        std::move(types)));
+  }
+
+  bool operator==(const Type& rhs) const override {
+    if (auto union_rhs = rhs.cast<UnionType>()) {
+      return types_ == union_rhs->types_;
+    }
+    return false;
+  }
+
+  at::ArrayRef<TypePtr> types() const {
+    return types_;
+  }
+
+  at::ArrayRef<TypePtr> containedTypes() const override {
+    return types();
+  }
+
+  // Check if a given TypePtr is an allowable member of this Union
+  bool can_hold_type(TypePtr& type) const {
+    return std::any_of(types_.begin(), types_.end(), [&](TypePtr union_contained_type){
```

Comment:
Nit: 

Check if `type` is `None` and use fast path of `can_hold_none`. Same for const reference variant of this method.